Caboodle/Permissions/Permissions.ios.cs

```diff
@@ -0,0 +1,122 @@
+            }
+
+            return Task.FromResult(PermissionStatus.Granted);
+        }
+
+        static Task<PermissionStatus> PlatformRequestAsync(PermissionType permission)
+        {
+            EnsureLocationPermissionDeclared(permission);
+
+            switch (permission)
+            {
+                case PermissionType.LocationWhenInUse:
+                    return RequestLocationAsync();
+            }
+
+            return Task.FromResult(PermissionStatus.Granted);
+        }
+
+        static void EnsureLocationPermissionDeclared(PermissionType permission)
+        {
+            var info = NSBundle.MainBundle.InfoDictionary;
+
+            if (permission == PermissionType.LocationWhenInUse)
+            {
+                if (!info.ContainsKey(new NSString("NSLocationWhenInUseUsageDescription")))
+                    throw new PermissionException("On iOS 8.0 and higher you must set either `NSLocationWhenInUseUsageDescription` or `NSLocationAlwaysUsageDescription` in your Info.plist file to enable Authorization Requests for Location updates!");
+            }
+        }
+
+        static PermissionStatus GetLocationStatus()
+        {
+            if (!CLLocationManager.LocationServicesEnabled)
+                return PermissionStatus.Disabled;
+
+            var status = CLLocationManager.Status;
+
+            if (UIDevice.CurrentDevice.CheckSystemVersion(8, 0))
+            {
+                switch (status)
+                {
+                    case CLAuthorizationStatus.AuthorizedAlways:
+                    case CLAuthorizationStatus.AuthorizedWhenInUse:
+                        return PermissionStatus.Granted;
+                    case CLAuthorizationStatus.Denied:
+                        return PermissionStatus.Denied;
+                    case CLAuthorizationStatus.Restricted:
+                        return PermissionStatus.Restricted;
+                    default:
+                        return PermissionStatus.Unknown;
+                }
+            }
+
+            switch (status)
+            {
+                case CLAuthorizationStatus.Authorized:
+                    return PermissionStatus.Granted;
+                case CLAuthorizationStatus.Denied:
+                    return PermissionStatus.Denied;
+                case CLAuthorizationStatus.Restricted:
+                    return PermissionStatus.Restricted;
+                default:
+                    return PermissionStatus.Unknown;
+            }
+        }
+
+        static Task<PermissionStatus> RequestLocationAsync()
+        {
+            if (!UIDevice.CurrentDevice.CheckSystemVersion(8, 0))
+                return Task.FromResult(PermissionStatus.Unknown);
+
+            var locationManager = new CLLocationManager();
+
+            EventHandler<CLAuthorizationChangedEventArgs> authCallback = null;
+            var tcs = new TaskCompletionSource<PermissionStatus>();
+
+            authCallback = (sender, e) =>
```

Comment:
Is it possible to use an inner function here so as to avoid this scary set-to-null-and-then-create-delegate?

Reply:
Also, it will not require a delegate to be created each time...